java/adapter/avro/src/test/java/org/apache/arrow/AvroToArrowIteratorTest.java

```diff
@@ -0,0 +1,152 @@
+
+  private AvroToArrowVectorIterator writeAndRead(Schema schema, List data) throws Exception {
+    File dataFile = TMP.newFile();
+
+    BinaryEncoder
+        encoder = new EncoderFactory().directBinaryEncoder(new FileOutputStream(dataFile), null);
+    DatumWriter writer = new GenericDatumWriter(schema);
+    BinaryDecoder
+        decoder = new DecoderFactory().directBinaryDecoder(new FileInputStream(dataFile), null);
+
+    for (Object value : data) {
+      writer.write(value, encoder);
+    }
+
+    return AvroToArrow.avroToArrowIterator(schema, decoder, allocator, /*targetBatchSize=*/3);
+  }
+
+  @Test
+  public void testStringType() throws Exception {
+    Schema schema = getSchema("test_primitive_string.avsc");
+    List<String> data = Arrays.asList("v1", "v2", "v3", "v4", "v5");
+
+    try (AvroToArrowVectorIterator iterator = writeAndRead(schema, data)) {
+      List<VectorSchemaRoot> roots = new ArrayList<>();
+      List<FieldVector> vectors = new ArrayList<>();
+      while (iterator.hasNext()) {
+        VectorSchemaRoot root = iterator.next();
+        FieldVector vector = root.getFieldVectors().get(0);
+        roots.add(root);
+        vectors.add(vector);
+      }
+      checkPrimitiveResult(data, vectors);
+      roots.forEach(root -> root.close());
+    }
+  }
+
+  @Test
+  public void testNullableStringType() throws Exception {
+    Schema schema = getSchema("test_nullable_string.avsc");
+
+    List<GenericRecord> data = new ArrayList<>();
+    List<String> expected = new ArrayList<>();
+    for (int i = 0; i < 5; i++) {
+      GenericRecord record = new GenericData.Record(schema);
+      String value = i % 2 == 0 ? "test" + i : null;
+      record.put(0, value);
+      expected.add(value);
+      data.add(record);
+    }
+
+    try (AvroToArrowVectorIterator iterator = writeAndRead(schema, data);) {
+      List<VectorSchemaRoot> roots = new ArrayList<>();
+      List<FieldVector> vectors = new ArrayList<>();
+      while (iterator.hasNext()) {
+        VectorSchemaRoot root = iterator.next();
+        FieldVector vector = root.getFieldVectors().get(0);
+        roots.add(root);
+        vectors.add(vector);
+      }
+      checkPrimitiveResult(expected, vectors);
+      roots.forEach(root -> root.close());
+    }
+
+  }
+
+  @Test
+  public void testRecordType() throws Exception {
+    Schema schema = getSchema("test_record.avsc");
+    ArrayList<GenericRecord> data = new ArrayList<>();
+    for (int i = 0; i < 5; i++) {
+      GenericRecord record = new GenericData.Record(schema);
+      record.put(0, "test" + i);
+      record.put(1, i);
+      record.put(2, i % 2 == 0);
+      data.add(record);
+    }
+
+    try (AvroToArrowVectorIterator iterator = writeAndRead(schema, data)) {
+      List<VectorSchemaRoot> roots = new ArrayList<>();
+      while (iterator.hasNext()) {
+        roots.add(iterator.next());
+      }
+      checkRecordResult(schema, data, roots);
```

Comment:
lets move these checks outside of the try with resources, so we can verify the iterator closing is independent of the roots closing.

Reply:
fixed.